pkg/descheduler/strategies/node_taint.go

```diff
@@ -0,0 +1,148 @@
+
+const (
+	TolerationOpExists v1.TolerationOperator = "Exists"
+	TolerationOpEqual  v1.TolerationOperator = "Equal"
+)
+
+// RemovePodsViolatingNodeTaints with elimination strategy
+func RemovePodsViolatingNodeTaints(ds *options.DeschedulerServer, strategy api.DeschedulerStrategy, policyGroupVersion string, nodes []*v1.Node, nodePodCount nodePodEvictedCount) {
+	if !strategy.Enabled {
+		return
+	}
+	deletePodsViolatingNodeTaints(ds.Client, policyGroupVersion, nodes, ds.DryRun, nodePodCount, ds.MaxNoOfPodsToEvictPerNode, ds.EvictLocalStoragePods)
+}
+
+// deletePodsViolatingNodeTaints evicts pods on the node which violate NoSchedule Taints on nodes
+func deletePodsViolatingNodeTaints(client clientset.Interface, policyGroupVersion string, nodes []*v1.Node, dryRun bool, nodePodCount nodePodEvictedCount, maxPodsToEvict int, evictLocalStoragePods bool) int {
+	podsEvicted := 0
+	for _, node := range nodes {
+		glog.V(1).Infof("Processing node: %#v\n", node.Name)
+		pods, err := podutil.ListEvictablePodsOnNode(client, node, evictLocalStoragePods)
+		if err != nil {
+			//no pods evicted as error encountered retrieving evictable Pods
+			return 0
+		}
+		totalPods := len(pods)
+		for i := 0; i < totalPods; i++ {
+			if maxPodsToEvict > 0 && nodePodCount[node]+1 > maxPodsToEvict {
+				break
+			}
+			if !checkPodsSatisfyTolerations(pods[i], node) {
+				success, err := evictions.EvictPod(client, pods[i], policyGroupVersion, dryRun)
+				if !success {
+					glog.Errorf("Error when evicting pod: %#v (%#v)\n", pods[i].Name, err)
+				} else {
+					nodePodCount[node]++
+					glog.V(1).Infof("Evicted pod: %#v (%#v)", pods[i].Name, err)
+				}
+			}
+		}
+		podsEvicted += nodePodCount[node]
+	}
+	return podsEvicted
+}
+
+// checkPodsSatisfyTolerations checks if the node's taints (NoSchedule) are still satisfied by pods' tolerations.
+func checkPodsSatisfyTolerations(pod *v1.Pod, node *v1.Node) bool {
+	tolerations := pod.Spec.Tolerations
+	taints := node.Spec.Taints
+	if len(taints) == 0 {
+		return true
+	}
+	noScheduleTaints := getNoScheduleTaints(taints)
+	allTolerated, _ := getMatchingTolerations(noScheduleTaints, tolerations)
+	if !allTolerated {
+		glog.V(2).Infof("Not all taints are tolerated after update for Pod %v on node %v", pod.Name, node.Name)
+		return false
+	}
+	return true
+}
+
+// getNoScheduleTaints return a slice of NoSchedule taints from the a slice of taints that it receives.
+func getNoScheduleTaints(taints []v1.Taint) []v1.Taint {
+
+	result := []v1.Taint{}
+	for i := range taints {
+		if taints[i].Effect == v1.TaintEffectNoSchedule {
+			result = append(result, taints[i])
+		}
+	}
+	return result
+
+}
+
+//toleratesTaint returns true if a toleration tolerates a taint, or false otherwise
+func toleratesTaint(toleration *v1.Toleration, taint *v1.Taint) bool {
+
+	if len(toleration.Effect) > 0 && toleration.Effect != taint.Effect {
+		return false
+	}
+
+	if len(toleration.Key) > 0 && toleration.Key != taint.Key {
+		return false
+	}
```

Comment:
could these be simplified to a single compounded `or` check?

Reply:
oh, apologies I misunderstood your previous comment. Thanks for your suggestion and code snippet to rectify this. Will update now :)